File: Sources/DEXT/TMFFBEncoder.cpp
#include "TMFFBEncoder.hpp"

namespace TMFFBEncoder {

static inline uint8_t pkID0(uint8_t slot) {
    return (uint8_t)(((uint16_t)slot * 0x1C + 0x1C) & 0xFF);
}

static inline uint8_t pkID1(uint8_t slot) {
    return (uint8_t)(((uint16_t)slot * 0x1C + 0x0E) & 0xFF);
}

static inline int8_t clampToInt8(int v) {
    if (v < -127) return -127;
    if (v > 127) return 127;
    return (int8_t)v;
}

size_t firstPacket(uint8_t slot, uint8_t code, const Envelope *env,
                   uint8_t *out, size_t outCap)
{
    if (outCap < 11) return 0;

    uint16_t attackTime = 0;
    int8_t   attackLvl  = 0;
    uint16_t fadeTime   = 0;
    int8_t   fadeLvl    = 0;
    if (env) {
        attackTime = (env->attackTime > 0xFFFF) ? 0xFFFF : (uint16_t)env->attackTime;
        attackLvl  = clampToInt8((int)env->attackLevel / 256);
        fadeTime   = (env->fadeTime > 0xFFFF) ? 0xFFFF : (uint16_t)env->fadeTime;
        fadeLvl    = clampToInt8((int)env->fadeLevel / 256);
    }

    out[0]  = 0xF0;
    out[1]  = pkID0(slot);
    out[2]  = code;
    out[3]  = (uint8_t)(attackTime & 0xFF);
    out[4]  = (uint8_t)(attackTime >> 8);
    out[5]  = (uint8_t)attackLvl;
    out[6]  = (uint8_t)(fadeTime & 0xFF);
    out[7]  = (uint8_t)(fadeTime >> 8);
    out[8]  = (uint8_t)fadeLvl;
    out[9]  = 0x46;
    out[10] = 0x54;
    return 11;
}

size_t updateConstantPacket(uint8_t slot, int8_t level,
                            uint8_t *out, size_t outCap)
{
    if (outCap < 4) return 0;
    out[0] = ClassConstant;
    out[1] = pkID1(slot);
    out[2] = 0x4F;
    out[3] = (uint8_t)level;
    return 4;
}

size_t updatePeriodicPacket(uint8_t slot, const PeriodicParams *p,
                            uint8_t *out, size_t outCap)
{
    if (!p || outCap < 8) return 0;
    int8_t mag    = clampToInt8((int)p->magnitude / 256);
    int8_t offset = clampToInt8((int)p->offset / 256);
    uint8_t phase = (uint8_t)((unsigned)p->phase / 256);
    uint16_t period = (p->period > 0xFFFF) ? 0xFFFF : (uint16_t)p->period;

    out[0] = ClassPeriodic;
    out[1] = pkID1(slot);
    out[2] = 0x4F;
    out[3] = (uint8_t)mag;
    out[4] = (uint8_t)offset;
    out[5] = phase;
    out[6] = (uint8_t)(period & 0xFF);
    out[7] = (uint8_t)(period >> 8);
    return 8;
}

size_t updateConditionPacket(uint8_t slot, const ConditionParams *p,
                             CommitCode commit,
                             uint8_t *out, size_t outCap)
{
    if (!p || outCap < 11) return 0;
    int satMax = (commit == CommitSpring) ? 0x54 : 0x64;
    int8_t rightCoeff = clampToInt8((int)p->positiveCoefficient * 100 / 0x7F00);
    int8_t leftCoeff  = clampToInt8((int)p->negativeCoefficient * 100 / 0x7F00);
    int rightSatI = (int)p->positiveSaturation * satMax / 0x7F00;
    int leftSatI  = (int)p->negativeSaturation * satMax / 0x7F00;
    if (rightSatI > satMax) rightSatI = satMax;
    if (leftSatI  > satMax) leftSatI  = satMax;
    if (rightSatI < 0) rightSatI = 0;
    if (leftSatI < 0)  leftSatI  = 0;

    int centerSigned = (int)p->centerOffset * 500 / 0x7F00;
    if (centerSigned < -500) centerSigned = -500;
    if (centerSigned >  500) centerSigned =  500;
    uint16_t center = (uint16_t)((int16_t)centerSigned);

    int deadI = (int)p->deadBand * 1000 / 0xFFFF;
    if (deadI > 1000) deadI = 1000;
    if (deadI < 0)    deadI = 0;
    uint16_t dead = (uint16_t)deadI;

    out[0]  = ClassCondition;
    out[1]  = pkID1(slot);
    out[2]  = 0x4F;
    out[3]  = (uint8_t)rightCoeff;
    out[4]  = (uint8_t)leftCoeff;
    out[5]  = (uint8_t)(center & 0xFF);
    out[6]  = (uint8_t)(center >> 8);
    out[7]  = (uint8_t)(dead & 0xFF);
    out[8]  = (uint8_t)(dead >> 8);
    out[9]  = (uint8_t)rightSatI;
    out[10] = (uint8_t)leftSatI;
    return 11;
}

size_t commitPacket(uint8_t slot, CommitCode commit, uint16_t durationMs,
                    uint8_t *out, size_t outCap)
{
    if (outCap < 15) return 0;
    uint16_t etype = (uint16_t)commit;
    out[0]  = 0xF0;
    out[1]  = slot;
    out[2]  = (uint8_t)(etype & 0xFF);
    out[3]  = (uint8_t)(etype >> 8);
    out[4]  = (uint8_t)(durationMs & 0xFF);
    out[5]  = (uint8_t)(durationMs >> 8);
    out[6]  = 0x00;
    out[7]  = 0x00;
    out[8]  = 0x00;
    out[9]  = pkID1(slot);
    out[10] = 0x00;
    out[11] = pkID0(slot);
    out[12] = 0x00;
    out[13] = 0x00;
    out[14] = 0x00;
    return 15;
}

size_t startEffectPacket(uint8_t slot, uint8_t repeats,
                         uint8_t *out, size_t outCap)
{
    if (outCap < 4) return 0;
    out[0] = 0x60;
    out[1] = slot;
    out[2] = ControlPlay;
    out[3] = repeats;
    return 4;
}

size_t stopEffectPacket(uint8_t slot, uint8_t *out, size_t outCap)
{
    if (outCap < 4) return 0;
    out[0] = 0x60;
    out[1] = slot;
    out[2] = ControlStop;
    out[3] = 0x00;
    return 4;
}

} // namespace TMFFBEncoder

File: Sources/Drivers/Thrustmaster/TMCommon/TMFFBEncoder.swift
#if canImport(WheelProtocol)
import WheelProtocol
#endif

public enum TMFFBEncoder {

    public static func encode(_ effect: NormalizedEffect) throws -> [USBPacket] {
        switch effect {
        case .constant(let slot, let magnitude, let duration, _, let env):
            return constantUpload(slot: slot,
                                  magnitude: magnitude,
                                  duration: duration,
                                  envelope: env)
        case .periodic(let slot, let kind, let p, let duration, let env):
            return periodicUpload(slot: slot,
                                  kind: kind,
                                  params: p,
                                  duration: duration,
                                  envelope: env)
        case .spring(let slot, let c):
            return conditionUpload(slot: slot, commit: .spring, params: c)
        case .damper(let slot, let c):
            return conditionUpload(slot: slot, commit: .damper, params: c)
        case .friction:
            throw DriverError.effectNotSupported(.friction)
        case .inertia:
            throw DriverError.effectNotSupported(.inertia)
        case .ramp:
            throw DriverError.effectNotSupported(.ramp)
        case .customForceData:
            throw DriverError.effectNotSupported(.customForceData)
        }
    }

    public static func startEffectPacket(slot: UInt8, repeats: UInt8 = 1) -> USBPacket {
        .interruptOut(endpoint: TMOpcode.interruptOutEndpoint,
                      bytes: [0x60, slot, TMOpcode.EffectControl.play.rawValue, repeats])
    }

    public static func stopEffectPacket(slot: UInt8) -> USBPacket {
        .interruptOut(endpoint: TMOpcode.interruptOutEndpoint,
                      bytes: [0x60, slot, TMOpcode.EffectControl.stop.rawValue, 0x00])
    }

    private static func constantUpload(slot: UInt8,
                                       magnitude: Int16,
                                       duration: UInt32,
                                       envelope: Envelope?) -> [USBPacket]
    {
        let level = Int8(truncatingIfNeeded: max(-127, min(127, Int(magnitude / 256))))
        let firstPkt = firstPacket(slot: slot,
                                   code: .constantPeriodic,
                                   envelope: envelope)
        let updatePkt = updateConstantPacket(slot: slot, level: level)
        let commitPkt = commitPacket(slot: slot,
                                     commit: .constant,
                                     durationMs: UInt16(min(duration, UInt32(UInt16.max))))
        return [firstPkt, updatePkt, commitPkt]
    }

    private static func periodicUpload(slot: UInt8,
                                       kind: EffectKind,
                                       params: PeriodicParams,
                                       duration: UInt32,
                                       envelope: Envelope?) -> [USBPacket]
    {
        let commit: TMOpcode.CommitCode
        switch kind {
        case .sinePeriodic:         commit = .sine
        case .squarePeriodic:       commit = .square
        case .trianglePeriodic:     commit = .triangle
        case .sawtoothUpPeriodic:   commit = .sawUp
        case .sawtoothDownPeriodic: commit = .sawDown
        default:                    commit = .sine
        }
        let firstPkt = firstPacket(slot: slot, code: .constantPeriodic, envelope: envelope)
        let updatePkt = updatePeriodicPacket(slot: slot, params: params)
        let commitPkt = commitPacket(slot: slot,
                                     commit: commit,
                                     durationMs: UInt16(min(duration, UInt32(UInt16.max))))
        return [firstPkt, updatePkt, commitPkt]
    }

    private static func conditionUpload(slot: UInt8,
                                        commit: TMOpcode.CommitCode,
                                        params: ConditionParams) -> [USBPacket]
    {
        let firstPkt = firstPacket(slot: slot, code: .condition, envelope: nil)
        let updatePkt = updateConditionPacket(slot: slot, params: params, commit: commit)
        let commitPkt = commitPacket(slot: slot, commit: commit, durationMs: 0)
        return [firstPkt, updatePkt, commitPkt]
    }

    private static func firstPacket(slot: UInt8,
                                    code: TMOpcode.FirstCode,
                                    envelope: Envelope?) -> USBPacket
    {
        let pkID0 = UInt8((UInt16(slot) &* 0x1C &+ 0x1C) & 0xFF)
        let attackTime  = UInt16(min(envelope?.attackTime ?? 0, UInt32(UInt16.max)))
        let attackLevel = UInt8(truncatingIfNeeded: Int(envelope?.attackLevel ?? 0) / 256)
        let fadeTime    = UInt16(min(envelope?.fadeTime ?? 0, UInt32(UInt16.max)))
        let fadeLevel   = UInt8(truncatingIfNeeded: Int(envelope?.fadeLevel ?? 0) / 256)
        let bytes: [UInt8] = [
            0xF0, pkID0, code.rawValue,
            UInt8(attackTime & 0xFF), UInt8(attackTime >> 8),
            attackLevel,
            UInt8(fadeTime & 0xFF), UInt8(fadeTime >> 8),
            fadeLevel,
            0x46, 0x54,
        ]
        return .interruptOut(endpoint: TMOpcode.interruptOutEndpoint, bytes: bytes)
    }

    private static func updateConstantPacket(slot: UInt8, level: Int8) -> USBPacket {
        let pkID1 = UInt8((UInt16(slot) &* 0x1C &+ 0x0E) & 0xFF)
        let bytes: [UInt8] = [
            TMOpcode.EffectClass.constant.rawValue,
            pkID1, 0x4F,
            UInt8(bitPattern: level),
        ]
        return .interruptOut(endpoint: TMOpcode.interruptOutEndpoint, bytes: bytes)
    }

    private static func updatePeriodicPacket(slot: UInt8, params: PeriodicParams) -> USBPacket {
        let pkID1 = UInt8((UInt16(slot) &* 0x1C &+ 0x0E) & 0xFF)
        let mag = Int8(truncatingIfNeeded: Int(params.magnitude) / 256)
        let off = Int8(truncatingIfNeeded: Int(params.offset) / 256)
        let phase = UInt8(truncatingIfNeeded: Int(params.phase) / 256)
        let period = UInt16(min(params.period, UInt32(UInt16.max)))
        let bytes: [UInt8] = [
            TMOpcode.EffectClass.periodic.rawValue,
            pkID1, 0x4F,
            UInt8(bitPattern: mag),
            UInt8(bitPattern: off),
            phase,
            UInt8(period & 0xFF), UInt8(period >> 8),
        ]
        return .interruptOut(endpoint: TMOpcode.interruptOutEndpoint, bytes: bytes)
    }

    private static func updateConditionPacket(slot: UInt8,
                                              params: ConditionParams,
                                              commit: TMOpcode.CommitCode) -> USBPacket
    {
        let pkID1 = UInt8((UInt16(slot) &* 0x1C &+ 0x0E) & 0xFF)
        let satMax: Int = (commit == .spring) ? 0x54 : 0x64
        let rightCoeff = Int8(truncatingIfNeeded: Int(params.positiveCoefficient) * 100 / 0x7F00)
        let leftCoeff  = Int8(truncatingIfNeeded: Int(params.negativeCoefficient) * 100 / 0x7F00)
        let rightSat = UInt8(min(Int(params.positiveSaturation) * satMax / 0x7F00, satMax))
        let leftSat  = UInt8(min(Int(params.negativeSaturation) * satMax / 0x7F00, satMax))
        let centerSigned = Int16(max(-500, min(500, Int(params.centerOffset) * 500 / 0x7F00)))
        let center = UInt16(bitPattern: centerSigned)
        let dead = UInt16(min(Int(params.deadBand) * 1000 / 0xFFFF, 1000))
        let bytes: [UInt8] = [
            TMOpcode.EffectClass.condition.rawValue,
            pkID1, 0x4F,
            UInt8(bitPattern: rightCoeff),
            UInt8(bitPattern: leftCoeff),
            UInt8(center & 0xFF), UInt8(center >> 8),
            UInt8(dead & 0xFF), UInt8(dead >> 8),
            rightSat, leftSat,
        ]
        return .interruptOut(endpoint: TMOpcode.interruptOutEndpoint, bytes: bytes)
    }

    private static func commitPacket(slot: UInt8,
                                     commit: TMOpcode.CommitCode,
                                     durationMs: UInt16) -> USBPacket
    {
        let pkID0 = UInt8((UInt16(slot) &* 0x1C &+ 0x1C) & 0xFF)
        let pkID1 = UInt8((UInt16(slot) &* 0x1C &+ 0x0E) & 0xFF)
        let etype = commit.rawValue
        let bytes: [UInt8] = [
            0xF0, slot,
            UInt8(etype & 0xFF), UInt8(etype >> 8),
            UInt8(durationMs & 0xFF), UInt8(durationMs >> 8),
            0x00, 0x00, 0x00,
            pkID1, 0x00,
            pkID0, 0x00,
            0x00, 0x00,
        ]
        return .interruptOut(endpoint: TMOpcode.interruptOutEndpoint, bytes: bytes)
    }
}
